File: forgeos-server/src/db/file-mutex.ts
/**
 * File-level mutex system for concurrent file lock management.
 *
 * Provides atomic file locking primitives that prevent two agents from
 * modifying the same workspace file concurrently. Backed by the
 * `file_locks` PostgreSQL table with a partial unique index on
 * `(file_path) WHERE released_at IS NULL` to guarantee mutual exclusion
 * at the database level.
 *
 * Core operations:
 * - {@link acquireFileLocks} — Lock files for a ticket (INSERT ON CONFLICT DO NOTHING)
 * - {@link checkFileConflicts} — Detect which files are held by other tickets
 * - {@link releaseFileLocks} — Release all active locks for a ticket
 *
 * All mutations are wrapped in transactions and emit audit events
 * (`FILE_LOCKED` / `FILE_UNLOCKED`) to the `events` table.
 *
 * @module db/file-mutex
 * @ticket TASK-FOS-04-003
 */

import { getPool } from './pool.js';
import { logger } from '../middleware/logging.js';
import type { FileLock } from '../types/index.js';

// ── Domain Error Types ───────────────────────────────────────────────────────

/**
 * Describes a single file conflict: the file path and the ticket holding it.
 */
export interface FileConflictDetail {
  /** Workspace-relative path of the conflicting file. */
  file_path: string;
  /** `ticket_id` of the ticket that holds the active lock. */
  locked_by_ticket: string;
  /** UUID of the agent holding the lock. `null` if unknown. */
  locked_by_agent: string | null;
  /** Hostname of the machine holding the lock. `null` if unknown. */
  locked_by_machine: string | null;
  /** ISO 8601 timestamp when the lock was acquired. */
  locked_at: string;
}

/**
 * Error thrown when file lock acquisition fails due to conflicts.
 *
 * Contains structured details of every conflicting file so callers can
 * report which files are locked by which tickets.
 */
export class FileConflictError extends Error {
  readonly code = 'FILE_CONFLICT' as const;
  readonly statusCode = 409;

  constructor(
    public readonly ticketId: string,
    public readonly conflicts: FileConflictDetail[],
  ) {
    const paths = conflicts.map((c) => c.file_path).join(', ');
    super(`File conflict for ticket ${ticketId}: files locked by other tickets — ${paths}`);
    this.name = 'FileConflictError';
  }
}

// ── Result Types ─────────────────────────────────────────────────────────────

/**
 * Result of a successful {@link acquireFileLocks} operation.
 */
export interface AcquireFileLocksResult {
  /** The `ticket_id` that acquired the locks. */
  ticket_id: string;
  /** Array of file paths that were successfully locked. */
  locked_files: string[];
  /** Number of files that were locked. */
  lock_count: number;
}

/**
 * Result of a successful {@link releaseFileLocks} operation.
 */
export interface ReleaseFileLocksResult {
  /** The `ticket_id` whose locks were released. */
  ticket_id: string;
  /** Array of file paths that were unlocked. */
  released_files: string[];
  /** Number of locks that were released. */
  release_count: number;
}

// ── Core Operations ──────────────────────────────────────────────────────────

/**
 * Check for file conflicts before attempting to acquire locks.
 *
 * Queries the `file_locks` table for active locks (released_at IS NULL)
 * on the given file paths that belong to a **different** ticket.
 *
 * @param ticketId - The ticket attempting to lock the files
 * @param filePaths - Workspace-relative file paths to check
 * @returns Array of conflict details; empty if no conflicts
 *
 * @example
 * ```ts
 * const conflicts = await checkFileConflicts('TASK-FOS-04-003', ['src/db/pool.ts']);
 * if (conflicts.length > 0) {
 *   throw new FileConflictError('TASK-FOS-04-003', conflicts);
 * }
 * ```
 */
export async function checkFileConflicts(
  ticketId: string,
  filePaths: string[],
): Promise<FileConflictDetail[]> {
  if (filePaths.length === 0) {
    return [];
  }

  const pool = getPool();

  const result = await pool.query<{
    file_path: string;
    ticket_id: string;
    locked_by: string | null;
    machine_id: string | null;
    locked_at: string;
  }>(
    `SELECT file_path, ticket_id, locked_by, machine_id, locked_at
     FROM file_locks
     WHERE file_path = ANY($1)
       AND released_at IS NULL
       AND ticket_id <> $2`,
    [filePaths, ticketId],
  );

  const conflicts: FileConflictDetail[] = result.rows.map((row) => ({
    file_path: row.file_path,
    locked_by_ticket: row.ticket_id,
    locked_by_agent: row.locked_by,
    locked_by_machine: row.machine_id,
    locked_at: new Date(row.locked_at).toISOString(),
  }));

  if (conflicts.length > 0) {
    logger.warn(
      {
        event: 'file_conflict_detected',
        ticketId,
        conflictCount: conflicts.length,
        conflictingFiles: conflicts.map((c) => c.file_path),
      },
      'File conflicts detected during lock check',
    );
  }

  return conflicts;
}

/**
 * Acquire file locks for a ticket's file paths.
 *
 * Uses `INSERT ... ON CONFLICT (file_path) WHERE released_at IS NULL DO NOTHING`
 * to atomically attempt to lock all files. If any file is already locked by a
 * different ticket, the conflicting INSERT is silently skipped.
 *
 * After the INSERT, the function compares the number of actually inserted rows
 * against the requested count. If any were skipped (conflict), it queries for
 * the conflicting locks and throws a {@link FileConflictError}.
 *
 * All operations run within a single transaction for atomicity: either all
 * locks are acquired or none are (on conflict, the transaction rolls back
 * the partial inserts).
 *
 * @param ticketId - The ticket acquiring the locks
 * @param filePaths - Workspace-relative file paths to lock
 * @param agentId - UUID of the agent acquiring the locks (nullable)
 * @param machineId - Hostname of the machine acquiring the locks (nullable)
 * @returns Result with the list of locked files
 * @throws {@link FileConflictError} if any file is already locked by another ticket
 *
 * @example
 * ```ts
 * const result = await acquireFileLocks(
 *   'TASK-FOS-04-003',
 *   ['src/db/file-mutex.ts'],
 *   'agent-uuid-123',
 *   'pop-os',
 * );
 * // result.locked_files === ['src/db/file-mutex.ts']
 * ```
 */
export async function acquireFileLocks(
  ticketId: string,
  filePaths: string[],
  agentId: string | null,
  machineId: string | null,
): Promise<AcquireFileLocksResult> {
  if (filePaths.length === 0) {
    return { ticket_id: ticketId, locked_files: [], lock_count: 0 };
  }

  const pool = getPool();
  const client = await pool.connect();

  try {
    await client.query('BEGIN');

    // Attempt to insert lock records for all file paths.
    // ON CONFLICT DO NOTHING silently skips files already locked.
    const insertResult = await client.query<{ file_path: string }>(
      `INSERT INTO file_locks (file_path, ticket_id, locked_by, machine_id)
       SELECT unnest($1::text[]), $2, $3, $4
       ON CONFLICT (file_path) WHERE released_at IS NULL
       DO NOTHING
       RETURNING file_path`,
      [filePaths, ticketId, agentId, machineId],
    );

    const lockedFiles = insertResult.rows.map((row) => row.file_path);
    const skippedCount = filePaths.length - lockedFiles.length;

    // If any inserts were skipped, there are conflicts — rollback and report.
    if (skippedCount > 0) {
      // Query the conflicting locks to provide detailed error information.
      const conflictResult = await client.query<{
        file_path: string;
        ticket_id: string;
        locked_by: string | null;
        machine_id: string | null;
        locked_at: string;
      }>(
        `SELECT file_path, ticket_id, locked_by, machine_id, locked_at
         FROM file_locks
         WHERE file_path = ANY($1)
           AND released_at IS NULL
           AND ticket_id <> $2`,
        [filePaths, ticketId],
      );

      const conflicts: FileConflictDetail[] = conflictResult.rows.map((row) => ({
        file_path: row.file_path,
        locked_by_ticket: row.ticket_id,
        locked_by_agent: row.locked_by,
        locked_by_machine: row.machine_id,
        locked_at: new Date(row.locked_at).toISOString(),
      }));

      await client.query('ROLLBACK');

      logger.warn(
        {
          event: 'file_lock_conflict',
          ticketId,
          requestedFiles: filePaths,
          conflicts: conflicts.map((c) => ({
            file: c.file_path,
            owner: c.locked_by_ticket,
          })),
        },
        'File lock acquisition failed due to conflicts',
      );

      throw new FileConflictError(ticketId, conflicts);
    }

    // Record FILE_LOCKED events for each acquired lock.
    for (const filePath of lockedFiles) {
      await client.query(
        `INSERT INTO events (ticket_id, event_type, agent_id, agent_name, machine_id, payload)
         VALUES ($1, 'FILE_LOCKED', $2, $3, $4, $5)`,
        [
          ticketId,
          agentId,
          agentId,
          machineId,
          JSON.stringify({ file_path: filePath }),
        ],
      );
    }

    await client.query('COMMIT');

    logger.info(
      {
        event: 'file_locks_acquired',
        ticketId,
        lockedFiles,
        lockCount: lockedFiles.length,
      },
      'File locks acquired successfully',
    );

    return {
      ticket_id: ticketId,
      locked_files: lockedFiles,
      lock_count: lockedFiles.length,
    };
  } catch (err) {
    // Rollback only if we haven't already (conflict path does its own rollback).
    if (!(err instanceof FileConflictError)) {
      await client.query('ROLLBACK').catch(() => {
        /* ignore rollback errors on already-failed connection */
      });
    }
    throw err;
  } finally {
    client.release();
  }
}

/**
 * Release all active file locks belonging to a ticket.
 *
 * Sets `released_at = NOW()` for every lock record where
 * `ticket_id` matches and `released_at IS NULL`. Records a
 * `FILE_UNLOCKED` audit event for each released lock.
 *
 * @param ticketId - The ticket whose locks should be released
 * @returns Result with the list of released file paths
 *
 * @example
 * ```ts
 * const result = await releaseFileLocks('TASK-FOS-04-003');
 * // result.released_files === ['src/db/file-mutex.ts']
 * ```
 */
export async function releaseFileLocks(
  ticketId: string,
): Promise<ReleaseFileLocksResult> {
  const pool = getPool();
  const client = await pool.connect();

  try {
    await client.query('BEGIN');

    // Release all active locks for the ticket, returning the released paths.
    const releaseResult = await client.query<{ file_path: string; locked_by: string | null; machine_id: string | null }>(
      `UPDATE file_locks
       SET released_at = NOW()
       WHERE ticket_id = $1
         AND released_at IS NULL
       RETURNING file_path, locked_by, machine_id`,
      [ticketId],
    );

    const releasedFiles = releaseResult.rows.map((row) => row.file_path);

    // Record FILE_UNLOCKED events for each released lock.
    for (const row of releaseResult.rows) {
      await client.query(
        `INSERT INTO events (ticket_id, event_type, agent_id, agent_name, machine_id, payload)
         VALUES ($1, 'FILE_UNLOCKED', $2, $3, $4, $5)`,
        [
          ticketId,
          row.locked_by,
          row.locked_by,
          row.machine_id,
          JSON.stringify({ file_path: row.file_path }),
        ],
      );
    }

    await client.query('COMMIT');

    if (releasedFiles.length > 0) {
      logger.info(
        {
          event: 'file_locks_released',
          ticketId,
          releasedFiles,
          releaseCount: releasedFiles.length,
        },
        'File locks released successfully',
      );
    } else {
      logger.debug(
        { event: 'file_locks_release_noop', ticketId },
        'No active file locks found for ticket',
      );
    }

    return {
      ticket_id: ticketId,
      released_files: releasedFiles,
      release_count: releasedFiles.length,
    };
  } catch (err) {
    await client.query('ROLLBACK').catch(() => {
      /* ignore rollback errors on already-failed connection */
    });
    throw err;
  } finally {
    client.release();
  }
}

/**
 * Get all active file locks for a specific ticket.
 *
 * @param ticketId - The ticket to query locks for
 * @returns Array of active FileLock records
 */
export async function getActiveLocksForTicket(
  ticketId: string,
): Promise<FileLock[]> {
  const pool = getPool();

  const result = await pool.query<FileLock>(
    `SELECT id, file_path, ticket_id, locked_by, machine_id, locked_at, released_at
     FROM file_locks
     WHERE ticket_id = $1
       AND released_at IS NULL
     ORDER BY locked_at ASC`,
    [ticketId],
  );

  return result.rows;
}

/**
 * Get the active lock for a specific file path, if any.
 *
 * @param filePath - Workspace-relative file path to check
 * @returns The active FileLock, or `null` if the file is unlocked
 */
export async function getActiveLockForFile(
  filePath: string,
): Promise<FileLock | null> {
  const pool = getPool();

  const result = await pool.query<FileLock>(
    `SELECT id, file_path, ticket_id, locked_by, machine_id, locked_at, released_at
     FROM file_locks
     WHERE file_path = $1
       AND released_at IS NULL`,
    [filePath],
  );

  return result.rows[0] ?? null;
}

File: mcp-server/src/mcp_server/locking/file_mutex.py
"""File-level advisory lock mutex using PostgreSQL advisory locks.

Provides :class:`FileMutex` for transaction-scoped file locking that prevents
two agents from modifying the same workspace file concurrently.  Uses
``pg_advisory_xact_lock`` / ``pg_try_advisory_xact_lock`` keyed on a
deterministic int64 hash of the file path.

Design decisions
----------------
* **Advisory locks, not row locks** --- advisory locks are lightweight,
  application-defined, and automatically released when the transaction
  ends (commit or rollback).  No cleanup required.
* **Deterministic hashing** --- file paths are hashed to int64 using a
  two-part CRC32-based scheme for reproducibility across sessions.
* **Observability via file_locks table** --- after advisory lock
  acquisition, a row is inserted into ``file_locks`` so operators can
  query active locks.  This is purely informational; the advisory lock
  is the authoritative mutex.
* **Dual locking modes** --- :meth:`FileMutex.acquire` blocks until the
  lock is available; :meth:`FileMutex.try_acquire` returns immediately
  with a success/failure indicator.

.. meta::
   :ticket: FORGEOS-BE007
"""

from __future__ import annotations

import struct
import zlib
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Protocol

if TYPE_CHECKING:
    from datetime import datetime

from mcp_server.observability import get_logger

logger = get_logger("locking.file_mutex")


# ---------------------------------------------------------------------------
# Hash function: file path -> int64 advisory lock key
# ---------------------------------------------------------------------------


def file_path_to_lock_key(file_path: str) -> int:
    """Convert a file path to a deterministic int64 advisory lock key.

    Uses CRC32 on the UTF-8 bytes of the normalized path to produce a
    32-bit hash, then packs it into a signed 64-bit integer.  The upper
    32 bits use a fixed namespace (``0x464F5247`` = "FORG") to avoid
    collisions with advisory locks used by other subsystems.

    Parameters
    ----------
    file_path : str
        Workspace-relative file path (e.g. ``"src/db/pool.py"``).

    Returns
    -------
    int
        Signed 64-bit integer suitable for ``pg_advisory_xact_lock``.

    Raises
    ------
    ValueError
        If *file_path* is empty.
    """
    if not file_path:
        raise ValueError("file_path must not be empty")

    # Normalize: strip leading/trailing whitespace and slashes.
    normalized = file_path.strip().strip("/")

    if not normalized:
        raise ValueError("file_path must not be empty")

    # CRC32 produces an unsigned 32-bit integer.
    path_hash = zlib.crc32(normalized.encode("utf-8")) & 0xFFFFFFFF

    # Pack namespace (upper 32 bits) + path hash (lower 32 bits) into
    # a signed int64, which is what PostgreSQL advisory locks expect.
    namespace = 0x464F5247  # ASCII "FORG"
    combined = (namespace << 32) | path_hash

    # Convert to signed int64 (PostgreSQL bigint is signed).
    return struct.unpack(">q", struct.pack(">Q", combined))[0]


# ---------------------------------------------------------------------------
# Domain types
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class FileLockRecord:
    """Observability record for an active file lock.

    Attributes
    ----------
    file_path : str
        Workspace-relative path that is locked.
    ticket_id : str
        Ticket that holds the lock.
    locked_by : str | None
        Agent UUID holding the lock, if known.
    machine_id : str | None
        Hostname of the machine holding the lock.
    locked_at : datetime
        Timestamp when the lock was acquired.
    """

    file_path: str
    ticket_id: str
    locked_by: str | None
    machine_id: str | None
    locked_at: datetime


@dataclass(frozen=True, slots=True)
class LockAcquireResult:
    """Result of a lock acquisition attempt.

    Attributes
    ----------
    acquired : bool
        ``True`` if the lock was successfully acquired.
    file_path : str
        The file path that was locked (or attempted).
    lock_key : int
        The int64 advisory lock key derived from the file path.
    ticket_id : str
        The ticket that requested the lock.
    """

    acquired: bool
    file_path: str
    lock_key: int
    ticket_id: str


class FileConflictError(Exception):
    """Raised when a file lock cannot be acquired due to an existing lock.

    Attributes
    ----------
    file_path : str
        The conflicting file path.
    ticket_id : str
        The ticket that attempted the lock.
    held_by_ticket : str | None
        The ticket currently holding the lock, if known.
    """

    def __init__(
        self,
        file_path: str,
        ticket_id: str,
        held_by_ticket: str | None = None,
    ) -> None:
        held = f" (held by {held_by_ticket})" if held_by_ticket else ""
        super().__init__(
            f"File lock conflict on '{file_path}' for ticket {ticket_id}{held}"
        )
        self.file_path = file_path
        self.ticket_id = ticket_id
        self.held_by_ticket = held_by_ticket


# ---------------------------------------------------------------------------
# Connection protocol (for dependency injection)
# ---------------------------------------------------------------------------


class ConnectionLike(Protocol):
    """Minimal async connection interface for advisory lock operations."""

    async def fetchval(self, query: str, *args: Any) -> Any: ...
    async def fetchrow(self, query: str, *args: Any) -> Any: ...
    async def execute(self, query: str, *args: Any) -> str: ...
    async def fetch(self, query: str, *args: Any) -> list[Any]: ...


# ---------------------------------------------------------------------------
# FileMutex -- core advisory lock operations
# ---------------------------------------------------------------------------


class FileMutex:
    """File-level mutual exclusion using PostgreSQL advisory locks.

    All operations require an active database connection that is inside
    a transaction.  The advisory lock is transaction-scoped -- it is
    automatically released when the transaction ends (commit or rollback).

    Parameters
    ----------
    conn : ConnectionLike
        An asyncpg connection (or compatible) inside an active transaction.
    """

    def __init__(self, conn: ConnectionLike) -> None:
        self._conn = conn

    async def acquire(
        self,
        file_path: str,
        ticket_id: str,
        agent_id: str | None = None,
        machine_id: str | None = None,
    ) -> LockAcquireResult:
        """Acquire a blocking advisory lock for a file path.

        Calls ``pg_advisory_xact_lock(key)`` which blocks until the lock
        is available.  After acquisition, inserts an observability record
        into the ``file_locks`` table.

        Parameters
        ----------
        file_path : str
            Workspace-relative file path to lock.
        ticket_id : str
            Ticket requesting the lock.
        agent_id : str | None
            Agent UUID, for observability tracking.
        machine_id : str | None
            Hostname, for observability tracking.

        Returns
        -------
        LockAcquireResult
            Result with ``acquired=True`` and the lock key.
        """
        lock_key = file_path_to_lock_key(file_path)

        logger.info(
            "Acquiring advisory lock",
            extra={
                "file_path": file_path,
                "lock_key": lock_key,
                "ticket_id": ticket_id,
            },
        )

        # pg_advisory_xact_lock blocks until acquired; returns void.
        await self._conn.execute(
            "SELECT pg_advisory_xact_lock($1)", lock_key
        )

        # Record in file_locks table for observability.
        await self._record_lock(file_path, ticket_id, agent_id, machine_id)

        logger.info(
            "Advisory lock acquired",
            extra={
                "file_path": file_path,
                "lock_key": lock_key,
                "ticket_id": ticket_id,
            },
        )

        return LockAcquireResult(
            acquired=True,
            file_path=file_path,
            lock_key=lock_key,
            ticket_id=ticket_id,
        )

    async def try_acquire(
        self,
        file_path: str,
        ticket_id: str,
        agent_id: str | None = None,
        machine_id: str | None = None,
    ) -> LockAcquireResult:
        """Attempt to acquire an advisory lock without blocking.

        Calls ``pg_try_advisory_xact_lock(key)`` which returns ``true``
        if the lock was acquired, ``false`` if it is held by another
        session.

        Parameters
        ----------
        file_path : str
            Workspace-relative file path to lock.
        ticket_id : str
            Ticket requesting the lock.
        agent_id : str | None
            Agent UUID, for observability tracking.
        machine_id : str | None
            Hostname, for observability tracking.

        Returns
        -------
        LockAcquireResult
            Result with ``acquired`` indicating success or failure.
        """
        lock_key = file_path_to_lock_key(file_path)

        logger.info(
            "Attempting non-blocking advisory lock",
            extra={
                "file_path": file_path,
                "lock_key": lock_key,
                "ticket_id": ticket_id,
            },
        )

        acquired: bool = await self._conn.fetchval(
            "SELECT pg_try_advisory_xact_lock($1)", lock_key
        )

        if acquired:
            await self._record_lock(file_path, ticket_id, agent_id, machine_id)
            logger.info(
                "Advisory lock acquired (try)",
                extra={
                    "file_path": file_path,
                    "lock_key": lock_key,
                    "ticket_id": ticket_id,
                },
            )
        else:
            logger.info(
                "Advisory lock not available (try)",
                extra={
                    "file_path": file_path,
                    "lock_key": lock_key,
                    "ticket_id": ticket_id,
                },
            )

        return LockAcquireResult(
            acquired=acquired,
            file_path=file_path,
            lock_key=lock_key,
            ticket_id=ticket_id,
        )

    async def release_ticket_locks(self, ticket_id: str) -> list[str]:
        """Release observability records for a ticket's file locks.

        Sets ``released_at = NOW()`` on all active ``file_locks`` rows
        for the given ticket.  The advisory locks themselves are released
        automatically when the transaction ends.

        Parameters
        ----------
        ticket_id : str
            Ticket whose lock records should be released.

        Returns
        -------
        list[str]
            File paths whose lock records were released.
        """
        rows = await self._conn.fetch(
            """UPDATE file_locks
               SET released_at = NOW()
               WHERE ticket_id = $1
                 AND released_at IS NULL
               RETURNING file_path""",
            ticket_id,
        )
        released = [row["file_path"] for row in rows]

        if released:
            logger.info(
                "Released file lock records",
                extra={
                    "ticket_id": ticket_id,
                    "released_files": released,
                    "count": len(released),
                },
            )
        else:
            logger.debug(
                "No active file lock records to release",
                extra={"ticket_id": ticket_id},
            )

        return released

    async def get_active_locks(self, ticket_id: str) -> list[FileLockRecord]:
        """Query active file lock records for a ticket.

        Parameters
        ----------
        ticket_id : str
            Ticket to query.

        Returns
        -------
        list[FileLockRecord]
            Active lock records for the ticket.
        """
        rows = await self._conn.fetch(
            """SELECT file_path, ticket_id, locked_by, machine_id, locked_at
               FROM file_locks
               WHERE ticket_id = $1
                 AND released_at IS NULL""",
            ticket_id,
        )
        return [
            FileLockRecord(
                file_path=row["file_path"],
                ticket_id=row["ticket_id"],
                locked_by=row["locked_by"],
                machine_id=row["machine_id"],
                locked_at=row["locked_at"],
            )
            for row in rows
        ]

    async def check_conflicts(
        self, file_paths: list[str], ticket_id: str
    ) -> list[FileLockRecord]:
        """Check for existing locks on file paths held by other tickets.

        Parameters
        ----------
        file_paths : list[str]
            File paths to check.
        ticket_id : str
            The requesting ticket (excluded from conflict results).

        Returns
        -------
        list[FileLockRecord]
            Lock records for files locked by other tickets.
        """
        if not file_paths:
            return []

        rows = await self._conn.fetch(
            """SELECT file_path, ticket_id, locked_by, machine_id, locked_at
               FROM file_locks
               WHERE file_path = ANY($1)
                 AND released_at IS NULL
                 AND ticket_id <> $2""",
            file_paths,
            ticket_id,
        )
        return [
            FileLockRecord(
                file_path=row["file_path"],
                ticket_id=row["ticket_id"],
                locked_by=row["locked_by"],
                machine_id=row["machine_id"],
                locked_at=row["locked_at"],
            )
            for row in rows
        ]

    # -- Private helpers ---------------------------------------------------

    async def _record_lock(
        self,
        file_path: str,
        ticket_id: str,
        agent_id: str | None,
        machine_id: str | None,
    ) -> None:
        """Insert an observability record into file_locks.

        Uses ``ON CONFLICT DO NOTHING`` on the partial unique index
        ``(file_path) WHERE released_at IS NULL`` to avoid duplicates
        when re-acquiring a lock the same ticket already holds.
        """
        await self._conn.execute(
            """INSERT INTO file_locks (file_path, ticket_id, locked_by, machine_id)
               VALUES ($1, $2, $3, $4)
               ON CONFLICT (file_path) WHERE released_at IS NULL
               DO NOTHING""",
            file_path,
            ticket_id,
            agent_id,
            machine_id,
        )
